src/thor/trace_route_action.cc

```diff
@@ -490,29 +504,76 @@ thor_worker_t::map_match(Api& request) {
+
+            // if its the first or last path info in the edge group then its been already trimmed
+            bool begin_trimmed = leg_begin == std::get<0>(edge_group);
+            bool end_trimmed = leg_end == std::get<1>(edge_group);
+            bool trivial_group = std::get<0>(edge_group) == std::get<1>(edge_group);
+            bool trivial_leg = leg_begin == leg_end;
+
+            // we need to scale the elapsed time of the current edge to undo what FormPath did
+            double begin_pct = begin_trimmed ? std::get<2>(edge_group)->distance_along : 0;
+            double end_pct = end_trimmed ? std::get<3>(edge_group)->distance_along : 1;
+            double begin_edge_scale = 1.0 / ((trivial_group ? end_pct : 1) - begin_pct);
+            double end_edge_scale = 1.0 / (end_pct - (trivial_group ? begin_pct : 0));
+
+            // we get the time up to the last edge before this begin edge if any. we also remove
+            // the turn cost at the begging of this edge if there is any
+            double trim_begin =
+                begin_trimmed ? 0 : std::prev(leg_begin)->elapsed_time + leg_begin->turn_cost;
+            // then we scale the elapsed time on the edge based on the distance along the edge but
+            // we need to scale that distance down in the case FormPath knew it was a partial edge
+            trim_begin += (leg_begin->elapsed_time - trim_begin) *
+                          (origin_match_result->distance_along - begin_pct) * begin_edge_scale;
+
+            // we get the time up to the last edge before this end edge if any. we also remove
+            // the turn cost at the begging of this edge if there is any
+            double trim_end = trivial_leg && begin_trimmed
+                                  ? 0
+                                  : std::prev(leg_end)->elapsed_time + leg_end->turn_cost;
+            // then we scale the elapsed time on the edge based on the distance along the edge but
+            // we need to scale that distance down in the case FormPath knew it was a partial edge
+            trim_end = (leg_end->elapsed_time - trim_end) *
+                       (end_pct - destination_match_result->distance_along) * end_edge_scale;
 
             // add a new leg to the current route
-            TripLegBuilder::Build(controller, matcher->graphreader(), mode_costing,
-                                  edges.begin() + last_edge_index, edges.begin() + i + 1,
-                                  *origin_location, *destination_location,
+            TripLegBuilder::Build(controller, matcher->graphreader(), mode_costing, leg_begin,
+                                  leg_end + 1, *origin_location, *destination_location,
                                   std::list<valhalla::Location>{}, *route->mutable_legs()->Add(),
-                                  interrupt, &m_temp_route_discontinuities, true);
+                                  interrupt, &m_temp_route_discontinuities, trim_begin, trim_end);
```

Comment:
the trick is we send an amount to trim from the beginning of the leg (to remove the offset previous legs have had on the the time) but also the partial elapsed time at the end (for all but the last leg)